hal/src/dma/controller.rs

```diff
@@ -0,0 +1,270 @@
+//! DMA controller (DMAC)
+
+use vcell::VolatileCell;
+
+/// (DMAC_BTCTRL) Source Address Increment Enable
+pub(super) const BTCTRL_SRCINC: u16 = 1 << 10;
+
+/// (DMAC_BTCTRL) Destination Address Increment Enable
+pub(super) const BTCTRL_DSTINC: u16 = 1 << 11;
+
+/// (DMAC_BTCTRL) Step Selection
+pub(super) const BTCTRL_STEPSEL: u16 = 1 << 12;
+
+/// (DMAC_BTCTRL) Beat Size Mask
+pub(super) const BTCTRL_BEATSIZE_MASK: u16 = 3 << BTCTRL_BEATSIZE_POS;
+
+/// (DMAC_BTCTRL) Beat Size Mask
+pub(super) const BTCTRL_BEATSIZE_POS: u16 = 8;
+
+/// (DMAC_BTCTRL) Step Size Mask
+pub(super) const BTCTRL_STEPSIZE_MASK: u16 = 7 << BTCTRL_STEPSIZE_POS;
+
+/// (DMAC_BTCTRL) Step Size Mask
+pub(super) const BTCTRL_STEPSIZE_POS: u16 = 13;
+
+/// (DMAC_BTCTRL) Descriptor Valid
+pub(super) const BTCTRL_VALID: u16 = 1;
+
+/// (DMAC_CHINTENCLR) Mask Register
+pub(super) const CHINTENCLR_MASK: u8 = 0x07;
+
+/// Channel Transfer Error Interrupt Enable
+pub(super) const CHINTENCLR_TERR: u8 = 1;
+
+/// Channel Transfer Complete Interrupt Enable
+pub(super) const CHINTENCLR_TCMPL: u8 = 1 << 1;
+
+/// Channel Suspend Interrupt Enable
+pub(super) const CHINTENCLR_SUSP: u8 = 1 << 2;
+
+/// (DMAC_CHINTENSET) Mask Register
+pub(super) const CHINTENSET_MASK: u8 = 0x07;
+
+/// Disable event output
+pub(super) const EVENT_OUTPUT_DISABLE: u16 = 0;
+
+/// Number of DMAC channels (SAMD21)
+#[cfg(not(feature = "samd51"))]
+pub(super) const NUM_CHANNELS: usize = 1;
+
+/// Number of DMAC channels (SAMD51)
+#[cfg(feature = "samd51")]
+pub(super) const NUM_CHANNELS: usize = 32;
+
+#[cfg(not(feature = "samd51"))]
+pub(super) const NVIC_PRIO_BITS: u16 = 2;
+
+#[cfg(feature = "samd51")]
+pub(super) const NVIC_PRIO_BITS: u16 = 3;
+
+/// Channel IDs
+pub(super) type ChannelId = u8;
+
+/// Compute BTCTRL value for beat size
+pub(super) fn btctrl_beatsize(size: u16) -> u16 {
+    (0x3 << 8) & (size << 8)
+}
+
+/// Compute BTCTRL value for step size
+pub(super) fn btctrl_stepsize(size: u16) -> u16 {
+    (0x7 << 13) & (size << 13)
+}
+
+/// Beat sizes
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub enum BeatSize {
+    /// 8-bit
+    Byte = 0,
+
+    /// 16-bit
+    HWord = 1,
+
+    /// 32-bit
+    Word = 2,
+}
+
+/// DMA descriptor
+#[repr(align(16))]
+pub struct Descriptor {
```

Comment:
Is this the best way to define the DMAC descriptor struct, or can it be defined via SVD somehow?